dropshot_endpoint/src/metadata.rs

```diff
@@ -239,6 +285,87 @@ impl ValidatedEndpointMetadata {
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum VersionRange {
+    All,
+    From(semver::Version),
+    Until(semver::Version),
+    FromUntil(semver::Version, semver::Version),
+}
+
+fn parse_semver(v: &syn::LitStr) -> syn::Result<semver::Version> {
+    v.value()
+        .parse::<semver::Version>()
+        .map_err(|e| {
+            syn::Error::new_spanned(v, format!("expected semver: {}", e))
+        })
+        .and_then(|s| {
+            if s.pre == semver::Prerelease::EMPTY {
+                Ok(s)
+            } else {
+                Err(syn::Error::new_spanned(
+                    v,
+                    String::from(
+                        "semver pre-release string is not supported here",
+                    ),
+                ))
+            }
+        })
+        .and_then(|s| {
+            if s.build == semver::BuildMetadata::EMPTY {
+                Ok(s)
+            } else {
+                Err(syn::Error::new_spanned(
+                    v,
+                    String::from("semver build metadata is not supported here"),
+                ))
+            }
+        })
+}
+
+impl syn::parse::Parse for VersionRange {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let lookahead = input.lookahead1();
+        if lookahead.peek(syn::Token![..]) {
+            let _ = input.parse::<syn::Token![..]>()?;
+            if input.is_empty() {
+                Ok(VersionRange::All)
+            } else {
+                let latest = input.parse::<syn::LitStr>()?;
+                let latest_semver = parse_semver(&latest)?;
+                Ok(VersionRange::Until(latest_semver))
+            }
+        } else {
+            let earliest = input.parse::<syn::LitStr>()?;
+            let earliest_semver = parse_semver(&earliest)?;
+            let _ = input.parse::<syn::Token![..]>()?;
+            let lookahead = input.lookahead1();
+            if lookahead.peek(syn::LitStr) {
+                let latest = input.parse::<syn::LitStr>()?;
+                let latest_semver = parse_semver(&latest)?;
+                if latest_semver < earliest_semver {
+                    let span: TokenStream = [
+                        TokenTree::from(earliest.token()),
+                        TokenTree::from(latest.token()),
+                    ]
+                    .into_iter()
+                    .collect();
+                    return Err(syn::Error::new_spanned(
+                        span,
```

Comment:
I think we want 
```suggestion
                        input.span(),
```

Maybe? In particular I'm thinking of this test:

dropshot/tests/fail/bad_version_backwards.stderr
```
error: semver range (from ... until ...) has the endpoints out of order
  --> tests/fail/bad_version_backwards.rs:17:16
   |
17 |     versions = "1.2.3".."1.2.2"
   |    
```

Should we be underlining both?

Reply:
This doesn't quite work:

```
error[E0277]: the trait bound `proc_macro2::Span: ToTokens` is not satisfied
   --> dropshot_endpoint/src/metadata.rs:354:25
    |
353 |                     return Err(syn::Error::new_spanned(
    |                                ----------------------- required by a bound introduced by this call
354 |                         input.span(),
    |                         ^^^^^^^^^^^^ the trait `ToTokens` is not implemented for `proc_macro2::Span`
    |
    = help: the following other types implement trait `ToTokens`:
              &'a T
              &'a mut T
              Abstract
              AndAnd
              AndEq
              AngleBracketedGenericArguments
              Arm
              As
            and 308 others
note: required by a bound in `syn::Error::new_spanned`
   --> /Users/dap/.cargo/registry/src/index.crates.io-6f17d22bba15001f/syn-2.0.79/src/error.rs:189:27
    |
189 |     pub fn new_spanned<T: ToTokens, U: Display>(tokens: T, message: U) -> Self {
    |                           ^^^^^^^^ required by this bound in `Error::new_spanned`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `dropshot_endpoint` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

I poked around a bit but did not immediately see how to turn that span into something that impls `ToTokens` (besides what the code is already doing).